libs/motor/inc/public/tfc/motors/virtual_motor.hpp

```diff
@@ -0,0 +1,194 @@
+#include <string>
+
+#include <mp-units/chrono.h>
+#include <mp-units/format.h>
+#include <mp-units/ostream.h>
+#include <mp-units/systems/isq/isq.h>
+#include <mp-units/systems/si/si.h>
+#include <boost/asio.hpp>
+
+#include <tfc/confman.hpp>
+#include <tfc/confman/observable.hpp>
+#include <tfc/motors/errors.hpp>
+#include <tfc/motors/impl.hpp>
+#include <tfc/utils/units_glaze_meta.hpp>
+
+namespace tfc::motor::types {
+using mp_units::QuantityOf;
+using tfc::confman::observable;
+using namespace mp_units::si::unit_symbols;
+namespace asio = boost::asio;
+namespace chrono = std::chrono;
+
+/**
+ * @brief Virtual motor class
+ * @details This class is used to simulate a motor, it does not actually control a motor.
+ * for use in testing use mock_motor in mock_motor.hpp. That motor can not be selected
+ * as a output of a running program and exposes a greater deal of internals.
+ */
+
+class virtual_motor {
+private:
+  struct config {
+    using impl = virtual_motor;
+    observable<std::string> name;
+    std::optional<mp_units::quantity<mp_units::si::milli<mp_units::si::metre> / mp_units::si::second>> nominal;
+
+    struct glaze {
+      using T = config;
+      static constexpr auto value = glz::object("name", &T::name, "nominal", &T::nominal);
+      static constexpr std::string_view name{ "printing_motor" };
+    };
+
+    auto operator==(const config&) const noexcept -> bool = default;
+  };
+
+public:
+  using config_t = config;
+
+  explicit virtual_motor(boost::asio::io_context& ctx, const config& config)
+      : ctx_(ctx), config_(config), logger_(config.name.value()) {
+    logger_.info("virtual_motor c-tor: {}", config.name.value());
+    config_.name.observe([this](std::string const& new_v, std::string const& old_v) {
+      logger_.warn("Printing motor name switched from: {}, to: {}! takes effect after this short message", old_v, new_v);
+      logger_ = logger::logger(new_v);
+    });
+  }
+
+  ~virtual_motor() {}
+
+  auto convey() -> std::error_code {
+    logger_.info("convey!");
+    running_ = true;
+    return {};
+  }
+
+  auto convey(QuantityOf<mp_units::isq::velocity> auto vel) -> std::error_code {
```

Comment:
maybe look into making one private `convey` function which could cover all public instances, less duplication. could use non type template parameters to activate or deactivate code